lib/SILOptimizer/Mandatory/TFDeabstraction.cpp

```diff
@@ -365,43 +373,69 @@ static BuiltinInst *simplifyOperands(BuiltinInst *inst, TFDeabstraction &TFDA) {
   // Predicate that returns true if the specified type is an address type for
   // a loadable (non-address-only) value.
   auto isLoadableAddressType = [&](SILType type) -> bool {
-    return type.isAddress() && type.isLoadable(inst->getModule());
+    return type.isAddress() && type.isLoadable(origInst->getModule());
   };
 
   // Predicate that returns true if an operand of the specified type should be
   // rewritten - either to load an address argument or expand a struct
   // parameter.
-  auto canSimplifyOperand = [&](SILType type) -> bool {
+  auto canSimplifyOperand = [&](SILType type,
+                                SILTensorOpInfo::OperandClass operandClass)
+      -> bool {
     return isLoadableAddressType(type) ||
-           getPrimitiveStructField(type.getASTType()) != nullptr;
+           getPrimitiveStructField(type.getASTType()) != nullptr ||
+           operandClass == SILTensorOpInfo::OperandClass::Out;
   };
 
   // If we don't have to change any operands, don't rewrite the builtin.
   bool mustChangeBuiltin = false;
-  for (auto &op : inst->getAllOperands()) {
-    if (canSimplifyOperand(op.get()->getType())) {
+  for (auto operandAndClass : zip(origInst->getAllOperands(),
+                                  opInfo.operandClasses)) {
+    auto operand = std::get<0>(operandAndClass).get();
+    auto operandClass = std::get<1>(operandAndClass);
+    if (canSimplifyOperand(operand->getType(), operandClass.second)) {
       mustChangeBuiltin = true;
       break;
     }
   }
 
-  if (!mustChangeBuiltin) return inst;
+  if (!mustChangeBuiltin) return origInst;
 
   // Mark the function as being mutated.
   TFDA.aboutToChangeFunction();
 
   // Okay, we do have to simplify something.  Scan through and rewrite operands.
-  SILBuilder B(inst);
+  SILBuilder B(origInst);
   SmallVector<SILValue, 8> operands;
-  for (auto &op : inst->getAllOperands()) {
-    auto operand = op.get();
+  SmallVector<std::pair<StringRef, SILTensorOpInfo::OperandClass>, 4>
```

Comment:
Have you considered not adding `$out` but just assume the very last (or first) input to be the indirect result? That won't require jointly iterating over all operands and operand classes, which minimizes the need for refactoring this loop.

Reply:
I feel like that would make things too implicit and hard to understand, and would add risk that someone mishandles the builtin somewhere if they don't realize that there can be an extra operand at the beginning or end.

Reply:
Ok. The down side is that it's making the code more complicated, and the only reason for refactoring this loop is to check for a single `$out` parameter... But it's okay for now.